plutus-tx/src/PlutusTx/DataMap.hs

```diff
@@ -0,0 +1,355 @@
+    go xs =
+      P.matchList
+        xs
+        True
+        ( \hd tl ->
+            let k = BI.fst hd
+             in if member' k tl then False else go tl
+        )
+
+{-# INLINEABLE all #-}
+all :: forall k a. (P.UnsafeFromData a) => (a -> Bool) -> Map k a -> Bool
+all p m = go (toBuiltinList m)
+  where
+    go :: BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData) -> Bool
+    go xs =
+      P.matchList
+        xs
+        True
+        ( \hd tl ->
+            let a = P.unsafeFromBuiltinData (BI.snd hd)
+             in if p a then go tl else False
+        )
+
+{-# INLINEABLE any #-}
+any :: forall k a. (P.UnsafeFromData a) => (a -> Bool) -> Map k a -> Bool
+any p m = go (toBuiltinList m)
+  where
+    go :: BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData) -> Bool
+    go xs =
+      P.matchList
+        xs
+        False
+        ( \hd tl ->
+            let a = P.unsafeFromBuiltinData (BI.snd hd)
+             in if p a then True else go tl
+        )
+
+{-# INLINEABLE union #-}
+
+-- | Combine two 'Map's.
+union ::
+  forall k a b.
+  (P.UnsafeFromData a, P.UnsafeFromData b, P.ToData a, P.ToData b) =>
+  Map k a ->
+  Map k b ->
+  Map k (These a b)
+union (toBuiltinList -> ls) (toBuiltinList -> rs) = unsafeFromBuiltinList res
+  where
+    ls' :: BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData)
+    ls' = go ls
+      where
+        go xs =
+          P.matchList
+            xs
+            nil
+            ( \hd tl ->
+                let k' = BI.fst hd
+                    v' = BI.snd hd
+                    v'' = case lookup' k' rs of
+                      Just r ->
+                        P.toBuiltinData
+                          ( These
+                              (P.unsafeFromBuiltinData v')
+                              (P.unsafeFromBuiltinData r) ::
+                              These a b
+                          )
+                      Nothing -> P.toBuiltinData (This (P.unsafeFromBuiltinData v') :: These a b)
+                 in BI.mkCons (BI.mkPairData k' v'') (go tl)
+            )
+
+    rs' :: BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData)
+    rs' = go rs
+      where
+        go xs =
+          P.matchList
+            xs
+            nil
+            ( \hd tl ->
+                let k' = BI.fst hd
+                    tl' = go tl
+                 in if member' k' ls
+                      then tl'
+                      else BI.mkCons hd tl'
+            )
+
+    res :: BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData)
+    res = go rs' ls'
+      where
+        go acc xs =
+          P.matchList
+            xs
+            acc
+            (\hd -> go (BI.mkCons hd acc))
+
+-- | Combine two 'Map's with the given combination function.
+unionWith ::
+  forall k a.
+  (P.UnsafeFromData a, P.ToData a) =>
+  (a -> a -> a) ->
+  Map k a ->
+  Map k a ->
+  Map k a
+unionWith f (toBuiltinList -> ls) (toBuiltinList -> rs) =
```

Comment:
Ah, this is the thing that lets users resolve clashes themselves, but it's less general than `union` because the values in both maps have to be of the same type.  I suppose one could generalise this by making it something like 
```
  forall k a.
  <suitable constraints> =>
  (a -> c) ->
  (b -> c) ->
  (a -> b -> c) ->
  Map k a ->
  Map k b ->
  Map k c
  ```
and then it would subsume `union`.